File: ruby/src/largest-rectangle-in-histogram.rb
# @param {Integer[]} heights
# @return {Integer}
def largest_rectangle_area(heights)
  stack = [[-1, 0]]
  max_area = 0
  (0..heights.size).each do |i|
    start = i
    while heights.fetch(i, 0) < stack.last[1]
      start, height = stack.pop
      area = height * (i - start)
      max_area = area if area > max_area
    end
    stack.push([start, heights[i]])
  end
  max_area
end

File: python3/largest-rectangle-in-histogram.py
import math


class Solution:
    """
    >>> Solution().largestRectangleArea([2,1,5,6,2,3])
    10
    >>> Solution().largestRectangleArea([2, 4])
    4
    >>> Solution().largestRectangleArea([2, 4, 2])
    6
    >>> Solution().largestRectangleArea([2, 1, 2])
    3
    >>> Solution().largestRectangleArea([2])
    2
    """

    def largestRectangleArea(self, heights: list[int]) -> int:
        stack = [(0, -1)]
        max_area = 0
        for i, height in enumerate(heights):
            start = None
            while len(stack) > 0 and height < stack[-1][0]:
                higher, start = stack.pop()
                max_area = max(max_area, higher * (i - start))
            start = i if start is None else start
            stack.append((height, start))
        while len(stack) > 0:
            higher, start = stack.pop()
            max_area = max(max_area, higher * (len(heights) - start))
        return max_area
